jeos-firstboot-enroll

```diff
@@ -12,197 +18,316 @@ FDE_SEAL_PCR_LIST="0,2,4,7,9"
+	if [ -e '/sys/class/tpm/tpm0' ]; then
+		if have_pcrlock && ! is_pcr_oracle; then
+			has_tpm2=lock
+		elif have_pcr_oracle; then
+			has_tpm2=oracle
+		fi
 	fi
-    fi
-
-   # For now seems that if a FIDO2 key is enrolled, it will take
-   # precedence over the TPM2 and the key will be asked to be present
-   # in subsequent boots.
-   if [ "$has_fido2" = '1' ] && [ -n "$has_tpm2" ]; then
-	local list=('FIDO2' 'FIDO2' 'TPM2' 'TPM2' 'none' $"Skip")
-	d --no-tags --default-item 'FIDO2' --menu $"Select unlock device" 0 0 "$(menuheight ${#list[@]})" "${list[@]}"
-	[ "$result" = 'FIDO2' ] && with_fido2=1
-	[ "$result" = 'TPM2' ] && with_tpm2="$has_tpm2"
-    elif [ "$has_fido2" ]; then
-	dialog $dialog_alternate_screen --backtitle "$PRETTY_NAME" --yesno $"Unlock encrypted disk via FIDO2 token?" 0 0 && with_fido2=1
-    elif [ -n "$has_tpm2" ]; then
-	dialog $dialog_alternate_screen --backtitle "$PRETTY_NAME" --yesno $"Unlock encrypted disk via TPM?" 0 0 && with_tpm2="$has_tpm2"
-    fi
-    return 0
+
+	while true; do
+		local list=()
+
+		if [ -z "$with_fido2" ] && [ -z "$with_tpm2" ] && [ -n "$has_fido2" ]; then
+			list+=('FIDO2' $'Enroll FIDO2 token')
+		fi
+		if [ -z "$with_tpm2" ] && [ -z "$with_fido2" ] && [ -n "$has_tpm2" ]; then
+			list+=('TPM2' $'Enroll TPM based token' 'TPM2_interactive' 'Enroll TPM based token with PIN')
+		fi
+		if [ -z "$crypt_pw" ]; then
+			if [ -n "$password" ]; then
+				list+=('root' $'Enroll root password')
+			fi
+			list+=('password' $'Enroll extra password')
+		fi
+		[ -n "$list" ] || break
+
+		list+=('done' $'Done')
+
+		d --no-tags --default-item "${list[0]}" --menu $"Disk Encryption" 0 0 "$(menuheight ${#list[@]})" "${list[@]}"
+		if [ "$result" = 'done' ]; then
+			if [ -z "$crypt_pw" ] && [ -z "$with_fido2" ] && [ -z "$with_tpm2" ] && [ -z "$is_jeos_config" ]; then
+				d_styled --yesno $"Neither password, TPM nor FIDO2 entrolled. Unlocking disk will only work with recovery key. Is this intended?" 0 0 || continue
+			fi
+			break;
+		elif [ "$result" = 'FIDO2' ]; then
+			with_fido2=1
+		elif [ "$result" = 'TPM2' ]; then
+			with_tpm2="$has_tpm2"
+		elif [ "$result" = 'TPM2_interactive' ]; then
+			while true; do
+				d --insecure --passwordbox  $"Enter new PIN (actually just passphrase)" 0 0
+				if [ -z "$result" ]; then
+					d_styled --yesno $"Retry?" 0 0 || break
+					continue
+				fi
+				crypt_tpm_pin="$result"
+				d --insecure --passwordbox  $"Confirm PIN" 0 0
+				[ "$crypt_tpm_pin" != "$result" ] || { with_tpm2="$has_tpm2"; break; }
+				d --msgbox $"PINs don't match. Try again" 0 0
+			done
+
+		elif [ "$result" = 'root' ]; then
+			crypt_pw="$password"
+		elif [ "$result" = 'password' ]; then
+			while true; do
+				d --insecure --passwordbox  $"Enter encryption password" 0 0
+				if [ -z "$result" ]; then
+					d --aspect 29 --msgbox $"No encryption password set. You can add more keys manually using cryptsetup." 0 0
```

Comment:
was commented before, but I would not recommend to use cryptsetup to the user, as should be systemd-cryptsetup now